### DISCUSSION TITLE
RFC 0024 - Implement The New Backoffice

### DISCUSSION THREAD

**Discussion by iOvergaard** [umbraco/rfcs#34] [RFC Discussions]
# Request for Comments: The new backoffice

### _Read the full [RFC document here](https://github.com/umbraco/rfcs/blob/main/cms/0024-implement-the-new-backoffice.md)._

[Back in March](https://umbraco.com/blog/future-proofing-the-backoffice-implementation/), we began talking about the implementation of the new backoffice. We also promised that there would be an RFC concerning the project, so today we are happy to announce that the final RFC (Request for comments) for the new backoffice has been published!

## How do I contribute?

You do not have to understand the proof of concept or any of the technologies to contribute. The most important thing is that we don’t want to miss anything, so everything goes in terms of clarifications, questions, suggestions, and so on.

Please do the following things if you want to contribute:

1. Read the RFC document [here](https://github.com/umbraco/rfcs/blob/main/cms/0024-implement-the-new-backoffice.md)
2. Check out the proof-of-concept application [here](https://github.com/umbraco/Umbraco.Backoffice.POC)
3. Please read and respect the [RFC Code of Conduct](https://github.com/umbraco/rfcs/blob/main/CODE_OF_CONDUCT.md)
4. Come back here and add comments down below 👇

## A little bit of history…

This RFC is the culmination of three other RFCs where we initially outlined the project, then went on to describe the need to build a UI library, and lastly, defined what the extension API would look like.

Read more here:

* [#0021 - Future-proofing the backoffice](https://github.com/umbraco/rfcs/blob/main/cms/0021-future-proofing-the-umbraco-backoffice.md)
* [#0022 - UI library](https://github.com/umbraco/rfcs/blob/main/cms/0022-ui-library.md)
* [#0023 - Defining the backoffice extension API ](https://github.com/umbraco/rfcs/blob/main/cms/0023-define-the-backoffice-extension-api.md)

This RFC is the final one for the backoffice, and you could say we are finally gluing everything together to present how to do the actual implementation of the new backoffice.

## The new RFC

This RFC consists of a document with the specifications and a proof-of-concept application you can download and try out for yourself.

### The RFC document

This document touches on a lot of subjects regarding the implementation of the new backoffice. A lot of choices need to be made for a web application, and we have tried to talk not only about those choices but also about what it means to the CMS as a whole.

### The proof-of-concept

We have decided to publish a [proof-of-concept application](https://github.com/umbraco/Umbraco.Backoffice.POC) along with the RFC document this time around. The application can be installed and run on your machine, and you can read the source code freely. The application demonstrates a few things on how to implement the backoffice, namely:

* How do we expect to build the web application
* How do we run a test suite
* What is the structure of the application
* How do we expect to handle contexts inside the application
* How do we expect extensions to interact with the backoffice and its contexts

Read the RFC in its entirety [here](https://github.com/umbraco/rfcs/blob/main/cms/0024-implement-the-new-backoffice.md) and make comments down below 👇

🗓 The RFC will be open for comments and considerations until Friday the 29th of July 2022 and will be evaluated (accepted/rejected) thereafter

**Comment by paulsterling** [comment]
Very impressive RFC! Agree this is the *best* direction - and fits nicely with the two prior RFCs.

Likes: backoffice as separate package from core CMS, Web Components, documentation. 

Concerns: TypeScript (that’s just a pill I’m going to have to swallow), number of dependencies (Lit, etc...). 

Looking forward to following along. And to be clear, I have no real concerns with the RFC, more just my own lack of experience with the tech stack proposed.

**Reply by iOvergaard** [reply]
Thank you for your sentiment, @paulsterling. 

TypeScript is useful to convey meaningful types, just like we do with JSON schemas today, and there might be convenient functions to use in your extensions as well, however, it is not obligatory to use in your own extensions where vanilla JavaScript is perfectly reasonable to use.

We will have to make sure that all of this is documented adequately for the helper tools made available to extension developers, as we talk about in the RFC 👍

**Reply by cssquirrel** [reply]
I'm not a huge fan of Lit because [insert Google's history of just dumping things], but I admit it does really help reduce web component boiler plate a ton, and without something like it vanilla components just take forever to code.

But genuinely I think this is going to be a very welcome change, and should keep the back office from growing stale due to a framework choice causing problems down the road.

(I for one *usually* like TypeScript, but I get why folks might find it tedious if unfamiliar.)

**Reply by markadrake** [reply]
I'm just reading through the RFC now, so please forgive me if I only add to the confusion above. 

However, my understanding from past conversations was that whether you used Lit or not, all UI components were simply web components. If one wanted to, it should be much easier to "integrate" with modern Angular, React, Vue, etc.? 

I'm thrilled to see this come to fruition. The back office has a lot of low-hanging fruit for substantial performance gains.

**Reply by iOvergaard** [reply]
@cssquirrel 

> I'm not a huge fan of Lit because [insert Google's history of just dumping things], but I admit it does really help reduce web component boiler plate a ton, and without something like it vanilla components just take forever to code.

YES, Google's history is unfortunate. That's kind of how we ended up having to replace AngularJS, all right. The key point here is **web components**. They won't go out of fashion anytime soon and whatever library (Lit) that glues stuff together can theoretically be replaced easily. Or perhaps Web Components evolve making Lit obsolete. This is an exciting future!

> (I for one _usually_ like TypeScript, but I get why folks might find it tedious if unfamiliar.)

Yes, well, usually I find strict _eslint_ rules to be more of a nuisance than TypeScript itself - you can really make life hard for developers with that. I just want to stress the point that going with TypeScript for backoffice development **does not** restrict extension developers in any way. In fact, the only requirement we may have for extension developers is that they ship their frontend code as an esmodule, but even that is undecided as of yet ¯\_(ツ)_/¯

**Reply by iOvergaard** [reply]
@markadrake 

> However, my understanding from past conversations was that whether you used Lit or not, all UI components were simply web components. If one wanted to, it should be much easier to "integrate" with modern Angular, React, Vue, etc.?

Absolutely correct. The runtime version of the backoffice UI is simply a bunch of Web Components/custom elements. Use them in any way you see fit with any library you want. That is the idea at least 🚀

**Comment by markadrake** [comment]
This may be out of scope, but I had hoped to see a local caching strategy for the Umbraco backoffice. Did any of these teams (internal or community) discuss utilizing IndexedDB as a layer of cache? This isn't a simple request, but done right could reduce the number of repetitious & redundant calls made during the editing experience.

**Low hanging fruit:** caching just a few things like the trees for content and media could end up saving editors a lot of time.

**Reply by iOvergaard** [reply]
@markadrake Yes, we want to do that, or at least lay the foundation for something like that in the future! 

The most important thing for the first release is to migrate the old backoffice to get out of the technical debt pit that we are in.

We want to investigate how we could prepare for this during development, e.g. making sure that we base the services around repositories so that a cache layer might be injected. But we excluded it from the RFC to keep the scope down and focus more on migration.

Some problems need to be solved for caching to work:

- Permissions: Currently permissions can be updated by _other users_ while you are logged in and a permission model is then included every time you fetch a content node for example so that the UI can react to the latest permission model. Obviously, permissions are validated on the server, but the best UX is to reflect the UI with the correct permissions.
- Large API responses: There are very few but very large API requests at the moment. We want to have more granulated requests with smaller payloads, so that we may decide better what can be cached. This is already in the works on the backend side of things.
- Probably more - this is just off the top of my head

**Reply by kjac** [reply]
@markadrake caching is an excellent thing to build on top of a great implementation; it does require a great implementation for starters, though, so caching doesn't a crucial corner stone for the entire application.

I'm certainly not bashing caching (rhyme unintentional), but I do believe that caching is an implementation detail to be worked out down the road, once a solid implementation foundation has been built. Therefore I think it's a bit premature to start defining caching strategies at this stage.

**Comment by argentini** [comment]
Before I dig in to the RFC, does it explain why unification of the technology for both platforms using something like Blazor wasn't chosen? It would be great to stay within the .NET world as much as possible.

**Reply by iOvergaard** [reply]
Hi @argentini

Thank you for your question.

I have just published [another answer](https://github.com/umbraco/rfcs/discussions/34#discussioncomment-2983514) to the more generic "why don't we choose my favorite framework" - please give that a read as well.

Specifically for Blazor, I can add that there are other reasons as well:

1. We want to extract the backoffice client as its own standalone web application not being bound to any specific backend
2. We do not want to have to choose between the latest hot framework from whatever vendor that may prove to be unstable or even be killed in the future
3. It makes sense to choose a stable backend technology for the backend (which will continue running on .NET) and a stable browser API for the frontend (something supported now and in the future for browsers)
4. We as a company must be able to find developers now and in the future who knows about specific aspects of web development, so in that way sticking solely to the Microsoft ecosystem greatly reduces the pool of developers available to us

I hope that answers most of your questions.

**Comment by iOvergaard** [comment]
### Why we are NOT choosing X framework

I just want to put an answer out there to the question: "Why don't you go with X framework?" as we have been asked that question many times.

Umbraco CMS is a piece of software that you install for perpetuity on a given computer/server. It is built with Microsoft's .NET technology, and the .NET Runtime can function on this machine for a prolonged period, eventually failing due to major updates of the OS, which can also be postponed.

The current backoffice is built using a series of frameworks and libraries to produce custom elements for the browser. The libraries also control other aspects of the browser such as navigation, animation, and hooks into various APIs of forms and validation.

All these tools depend entirely on the browser of the visitor at any given time - now and in the future. All these browsers exist in and consist of a dynamic environment, and that is *bound to change*. This means that, even though you may be able to install Umbraco CMS on your server and be ensured that the backend keeps working, there are no guarantees that the backoffice frontend will work 5 years from now, simply due to the nature of browsers and their maintainers.

Security is another issue; any external library that we install must be maintained and updated regularly whenever a new CVE is published. This is more easily done on an auto-updated product such as a SaaS product, but with Umbraco CMS being an installed product, we do not have that luxury. Umbraco 7 was released in 2013 using the new-at-the-time AngularJS framework. This framework is now considered end-of-life as of January 2022 and will no longer receive updates. We can and we will build a new backoffice, but there are a lot of old installations out there essentially risking their safety due to a choice (albeit a good one at the time) we made 8 years ago.

**We can do better than that.**

JavaScript frameworks are all doing essentially the same thing: Producing custom elements with syntactic sugar, and also added benefits of having built-in routers, state machines, and stores. The latter can easily be abstracted into a series of minor libraries (if not already supported by the browser). The former (custom elements) now exist as a native browser API known as [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components). Many frameworks also now support exporting their components as Web Components ensuring, that all components can communicate through a common API.

Building our software with Web Components - a native web standard - ensures that our software works and will keep working for at least a longer period than any external library will. The APIs will keep getting updated along with the browser itself, patching any security vulnerability along the way.

So why are we **not** choosing a framework? Because we do not have to 🚀

**Reply by iOvergaard** [reply]
Some great resources on Web Components:

[Build and publish your components as isolated modules](https://webcomponents.dev/)

[All the ways to make a Web Component using your favorite framework](https://webcomponents.dev/blog/all-the-ways-to-make-a-web-component/)

[Umbraco's favorite library: Lit](https://lit.dev/playground/)

**Comment by argentini** [comment]
Understood, and thanks for responding. If one of your main goals is to
stick to web standards and not rely on frameworks, then your choices make
sense to me. I don’t know that I agree with you on this versus keeping the
developer stack more unified for the .NET developers you guys already
serve, but I also understand that asking five developers for a technical
strategy will almost always yield five different answers :)

I wish you guys luck and am anxious to see the fruits of your labors :)

M.

**Reply by iOvergaard** [reply]
Thank you so much and thanks for the feedback!

**Comment by DanDiplo** [comment]
If you're not using any form of framework, then how are you going to deal with things like data binding, routing, filtering etc. Basically all the features that frameworks provide that handle this? For instance, currently AngularJs makes filtering of data in real-time and two-way binding very simple to achieve.

Won't you inevitably end up writing your own framework? Then you'll have to document it so people extending the back-office don't keep reinventing the wheel. This is all non-trivial. Then you'll have to maintain it, fix it etc. before you can even start doing the "proper" work. And if you don't, then people will just start throwing in their own favourite frameworks to packages, which will make the back-office bloated. How are you going to avoid that? 

I would have thought picking an existing lightweight framework to "glue" the components together and give people something common to use would be a better idea than rolling everything yourself.

**Reply by iOvergaard** [reply]
Hi @DanDiplo 

Thanks for your questions!

**Frameworks**
While we state we are not choosing a framework, it does not mean that we are not using other tools and libraries.

I believe this is a matter of definition. We are not rolling our own, but sticking to the Web Components API. In order to do so, we will be using the [Lit library](https://lit.dev/) to glue everything together. We will be pushing this library towards extension developers as well.

Incidentally, Web Components and Lit take care of most of the stuff we know from AngularJS, such as the concept of "components" and data binding. Running through a data set is fairly easy with Lit's template interpolation using native javascript to accomplish this in a much cleaner way than AngularJS did. Two-way data binding is, however, something that only AngularJS supported and is not something, you would use today. Instead, the flow would be: input = data, output = events.

```mermaid
flowchart
ComponentA -- data-binding --> ComponentB
ComponentB -- eventEmitter<Data> --> ComponentA
```

**Routing**
We are currently scoping out a library called [router-slot](https://www.npmjs.com/package/router-slot) which is included in the [proof-of-concept](https://github.com/umbraco/Umbraco.Backoffice.POC) for this project. It supports everything we know from the AngularJS router and more.

Please see [my other answer](https://github.com/umbraco/rfcs/discussions/34#discussioncomment-2983514) for more in-depth about frameworks.

**Reply by DanDiplo** [reply]
Hi @iOvergaard  Thanks for the helpful response. I'm glad you are leveraging existing tools, as writing everything from scratch sounds like a large endeavour!

I guess my main concern would be for people extending the back office. In my experience, the vast majority of people who do this are primarily backend developers with some front-end knowledge. For this audience, you want this to be as simple as possible. My perception is that few people are currently using web components, or any of the other libraries. So it will be imperative to have extensive documentation (something that Umbraco has been poor on in the past). 

But I also don't think "amateur" package developers are going to want to have to learn numerous new technologies that are pretty specific to Umbraco. I mean, I'd be happy to learn (say) `View.js` _if_ Umbraco used it, as it is something I could use outside of Umbraco.  But if you are using a load of minor libraries, then I'd be less interested in spending my free-time learning them. There are many of us who support Umbraco with our free-time to help build packages for the community. We're not full time front-end developers, we're often .NET devs who want a simple and tried-and-tested way of building what are often quite complex applications. So I hope you consider that. Thanks!

**Reply by kjac** [reply]
Hi @DanDiplo :)

I am quite sympathetic to your concerns. I share them to a large extend. One of my main motivations for joining the backoffice team was indeed to sound the voice of the backend developers, to ensure that they will still be able to create packages without having to become fully-fledged front-end developers first.

I believe most of us taught ourselves AngularJS back in the early V7 days, for the sole reason of being able to extend the backoffice and maybe contribute to the core. Eventually we might even have used those skills on customer projects as well, before AngularJS became sufficiently obsolete in the world of front-end. Moving to a new paradigm will undoubtedly mean having to learn new things, more-so if one wants to contribute code directly to the CMS. My take on this is that the learning curve will be far less steep this time around.

As for developing packages, you won't *have* to go the Lit way. You will be able to bring in your own frameworks and libraries and still work with the new backoffice components. Of course, if you (like me) don't know any other framework than AngularJS, you might need to learn a new one - I honestly don't know if AngularJS works with web components. On the other hand, you could also choose to go entirely vanilla JS and things would still work. Indeed, the RFC states there will be a starting point (template or something along those lines) for people not wanting to start out with Lit and TypeScript.

I guess what I'm trying to say here is: No-one wants to cut off the existing eco-system of contributors and package developers. There will be new things to learn, but it is in everyone's best interest to keep the learning curve as shallow as at all possible, and to include ample documentation moving forward.

**Reply by DanDiplo** [reply]
Hey @kjac - I think I'm both reassured and worried by what you say :)  I'm reassured that you're on the team and I definitely trust your judgement. I'm not opposed to learning new stuff (as you say, I learnt AngularJS for Umbraco) - it can be fun - but I know modern front-end can be quite daunting nowadays. I guess I just worry that if you don't use a single, well-documented framework, but pick a lot of different more obscure libraries, that makes it more difficult to pick-up.

But the bit that really worries me is the, "You will be able to bring in your own frameworks and libraries and still work with the new backoffice components." I really don't see this as a good thing at all. If developer A uses VueJs, developer B includes React, developer C includes a different version of VueJs it will end-up with a bloated back-office. Not to mention it will be a security nightmare when any of these imported frameworks require updating. 

But that isn't my only concern. If there isn't a common framework, then this will fracture the community. I *want* Umbraco to say, "You will use Framework X" and then _every_ single developer will use it. That helps unify developers together with a common shared language. It makes documentation much simpler, as there is only one way of doing things. It makes learner from other peoples' packages simpler, as you can guarantee they will be doing it the same way as you and everyone else. It makes helping each other simpler as we are all using the same shared framework and can learn together. 

But if everyone is using their own pet framework, it just makes it much harder to learn - be it learning from source code, documentation or from blog posts, articles, forum posts etc. So I want the backoffice team to define what is used and everyone be "forced" to use it. I don't see a free-for-all as being a good thing at all.

**Reply by kjac** [reply]
Hi @DanDiplo. Valid points indeed. 

We aren't choosing a framework because we don't want to have to rewrite the entire backoffice again when the chosen framework becomes obsolete, either due to end-of-life or lack of adoption. In other words, we want to utilize framework agnostic standards rather than framework specific features. This is where web components come into play. Web components encapsulate the underlying technology and expose an API that can be utilized by any consumer. Among other things, this prevents the inter-project dependencies (explicit and implicit) that are the main reason for breakage today.

I am entirely convinced that most package developers will go along the same path, using Lit and TypeScript and web components. Umbraco surely will be pushing it, precisely for the reasons you mention above.

The point of all this is .. being framework agnostic is not a goal in itself, but it is going to help ensure the future of the new backoffice and prevent yet another massive rewrite. It doesn't mean you _should_ bring in your own framework, but it does means you _can_ if you will. I trust the package developers to only do so if it is really, really necessary 😄

**Reply by Jogai** [reply]
>  I mean, I'd be happy to learn (say) View.js if Umbraco used it

@DanDiplo did you mean VueJs?

> we don't want to have to rewrite the entire backoffice again when the chosen framework becomes obsolete, either due to end-of-life or lack of adoption.

Picking one that has a high adoption rate already will prevent that.

> I trust the package developers to only do so if it is really, really necessary

Given the choice developers will stick to what they know, so its a given that more frameworks will be used in the future.

Maybe it helps a bit if there are examples for the lightweight variants like vue-petite etc. 

Not saying its the wrong choice, just highlighting some other viewpoints.

Also, did you evaluate FAST, the webcomponent framework from microsoft? Since a lot of dotnet devs will trust a choice like that and its already integrated with the various popular front-end frameworks out there: https://www.fast.design/docs/integrations/introduction

**Reply by iOvergaard** [reply]
@Jogai 

> Did you evaluate FAST

Are you thinking about the fast-components here as standard UI elements, or the fast-element to build new UI elements?

We are going to leverage the [Umbraco UI Library](https://github.com/umbraco/Umbraco.UI) to build the UI, which is a set of web components designed specifically for usage in the backoffice and other Umbraco products.

> Given the choice developers will stick to what they know, so its a given that more frameworks will be used in the future.

Developers today are already adopting a platitude of libraries and frameworks in the backoffice to build their packages, unfortunately. The nature of such an open ecosystem (as the Umbraco backoffice) allowing any arbitrary HTML/javascript to execute makes it possible to both customize to your heart's extent and abuse the system.

We want to set a good example by building our own documentation with a limited set of tools, e.g. TypeScript and Lit, and shipping a package-helper-tool to help developers get started building packages/extensions for the Umbraco backoffice. Hopefully, this will calm down the waters regarding having to "learn something new".

Our hope is that people will adopt the Web Components standard and get to learn, that you can [build Web Components with many different frameworks](https://webcomponents.dev/blog/all-the-ways-to-make-a-web-component/) without having to ship the entire runtime of them as well, so you can in many cases stick to what you know.

In any case, just as today, whatever the actual backoffice is being built with will have no effect on what extensions are built with. The runtime that extension developers work up against will be pure javascript and Web Components. The source code of the backoffice will not be shipped with the runtime version of Umbraco and any libraries we have chosen will not be leaking outside our code due to encapsulation, and we will also be running the runtime code through a tree-shaker which will leave only the functions that we ourselves use in the backoffice.

Therefore extension developers will have to bring their own libraries and tools to build their extensions (or just build them with vanilla javascript). We **will** be shipping a package that makes it easy to request a service/context from upstream, though, and we are experimenting with import maps to open up for sharing dependencies, e.g. Lit or SignalR or something else that we might use in the backoffice, so extensions at least won't have to include dependencies multiple times.

**Reply by Jogai** [reply]
I meant fast-element, because I believe that's an alternative to lit. (Again, nothing wrong with Lit, just curious if options from microsoft were considered).

> We want to set a good example by building our own documentation with a limited set of tools, e.g. TypeScript and Lit

Thats important to have indeed!

> without having to ship the entire runtime of them as well

Some of the runtime is shipped, depending on how good the treeshaking works.

**Reply by iOvergaard** [reply]
> > without having to ship the entire runtime of them as well

> Some of the runtime is shipped, depending on how good the treeshaking works.

Yes, you are right. [This blog post](https://webcomponents.dev/blog/all-the-ways-to-make-a-web-component/) contains a comparison between different frameworks and their bundle size to export a Web Component. Lit is something like 5 KB and Svelte ~2 KB for example.

By the way, we see it as acceptable if extensions to the backoffice have a 5 KB runtime overhead, but as I said, maybe some of it can be mitigated by [import maps](https://github.com/WICG/import-maps), which is still an experimental technique.

**Reply by Jogai** [reply]
That's certainly acceptable. And overall I think a major improvement in general and especially coming from angularjs

**Comment by biapar** [comment]
React?

**Reply by iOvergaard** [reply]
Hi @biapar - did you intend to write this as a reply to one of the other answers?

**Reply by biapar** [reply]
Hi, yes. Will be always Angular the Umbraco BackOffice?

**Reply by kjac** [reply]
Hi @biapar. The RFC proposes moving away from AngularJS in favor of a web component based architecture. Lit will be used for convenience when building the web components, but the resulting backoffice will be framework agnostic.

**Reply by biapar** [reply]
Ah ok. I use IBM Carbon React that used Lit Inside.

**Reply by iOvergaard** [reply]
Great, @biapar. React also allows you to build Web Components. [Check out this example](https://studio.webcomponents.dev/edit/MyduQavKRFmGHWvK91Z5?pm=1) :)

**Comment by argentini** [comment]
This is why I asked about Blazor. It's built-in, and is a .NET/C# technology. We've built whole apps with it, in production, and it's great.

**Reply by iOvergaard** [reply]
We don't want to have the backoffice being dependent on a web server in the background to run, which leaves out Blazor server-side mode. And in my opinion, Blazor WebAssembly mode is not desirable in a dynamic runtime mode and would make it hard for extension developers to interface with our services unless you rebuild the backoffice every time you install a new extension.

**Reply by argentini** [reply]
The front-end is always going to be dependent on a website, otherwise it won't be able to do anything anyway. And you still have to build the project every time you add packages since they're all in nuget now. So based on what I know at this point, Blazor Server is still a good option.

**Comment by tiffy74** [comment]
There's a number of places I could make this comment related to the new backoffice.

So first want to say that I totally support the idea of a web-component based backoffice - which I've been a fan of for some time and also presents some great opportunities.

As someone who has engaged recently with a lot of elements with the backoffice and how it's constructed - the idea of just a general wholescale rewrite is very appealing.  When I've been diving into the code of the backoffice I generally spot a lot of inconsistencies and areas where things have been pieced together primarily because they 'work' to some extent but haven't been consistently labelled or applied in the same way.  Take for example the different sections from Content through to Settings - inspecting the left hand column and reviewing the code really highlights some of these inconsistencies where components that look the same and do the same thing - are not coded in the same way.

So this approach is really most welcome...but..

What I would like to see is more methodology in the approach.  By this I mean adopting a set of principles on how the components and hence backoffice is written.  I've already seen some areas that to me, represent problems going forward.  I'm a huge fan of BEM for example which represents a method of providing a descriptors to a components respective parts and if done well - how it fits together - something that really makes for improving the experience of trying to identify how things are scoped and put together.  I'm not arguing necessarily for BEM - but I am suggesting a similar approach be adopted.  I totally support the idea of avoiding frameworks - but methodology is distinctly different to framework (although often some can conflate the two).

Lastly putting my accessibility hat on - and very much related to the previous point - everything needs to be built according to accessibility principles.  This will impact upon the direction you may wish to take with areas - and will require some real thinking 'outside the box' - but at the same time it represents a real opportunity to put Umbraco ahead of the pack.  The point being - and something that after chatting with some people who just don't 'realise' the ambition - is that the whole backoffice should be usable.. not just the editor experience.. but everything.  Being blind or having some other impairment - shouldn't be something that prevents you from being a developer - and hence we need to remove anything that may represent a barrier to that ambition :)

**Reply by iOvergaard** [reply]
Hi @tiffy74 

Thank you so much for your feedback!

Accessibility is top of mind for us as we proceed with the project. We will also be involving the Accessibility Community Team along the way ☺️ 

Regarding methodology. I hear what you are saying. We are currently not far enough to put any method into the RFC, but we will work something out that fits our needs. Working with Web Components and shadow DOM has its ups and downs in this regard; with the isolation of CSS we do not have to worry too much about the problems that something like BEM fixes, but naturally, we have to stick to some guidelines. One thing we have experimented with is splitting components into container components (with business logic) and view components, which allows for much easier testing access and UI documentation. We are very much open to suggestions in this area.

**Comment by PascalIcatt** [comment]
I would like to give my point of view regarding security.
Security requirements by governments, health industry standards and also national recommendations like the Dutch National Cyber Security Center https://www.ncsc.nl/ are becoming increasingly important.

I would like to request to research the possibilities of having the new backoffice fully Content-Security-Policy compliant.
And with that I mean to disallow inline styles and scripts and not use 'unsafe inline' and 'unsafe-eval'
(I have a draft version of the new NCSC standard that will be released soon, and this will also be included)

If we can do this, then Umbraco will be the first CMS (99% sure) that can be used for even the highest security requirements and this will be of huge benefit to the Cloud and Heartcore services as well.

Google paper (and bible)
[CSP Is Dead, Long Live CSP! On the Insecurity of Whitelists and the Future of Content Security Policy](https://research.google/pubs/pub45542/)

**Reply by kjac** [reply]
@PascalIcatt this is super interesting 👍 and super awesome to raise awareness! Would you happen to know if web component isolation on its own helps along this way, or if we should be doing additional things to achieve this? 

(Disclaimer: I'm vacationing and didn't dig into the "bible")

**Reply by iOvergaard** [reply]
Hi @PascalIcatt 

Thanks for bringing this to our attention. I'm pretty sure you can already set a pretty strict CSP header on whatever web server that hosts Umbraco/the backoffice, but "unsafe inline" in particular is probably not going to work on the current backoffice.

Using new tech for the frontend and separating the server and client with the new backoffice allows us to avoid doing stuff inline with the new backoffice.

I will make sure that we add an item to the backlog to research the possibilities of different CSP values.

**Comment by hemraker** [comment]
[ The RFC](https://github.com/umbraco/rfcs/blob/main/cms/0024-implement-the-new-backoffice.md) has been accepted and merged 🎉 Thank you all for the input and feedback - it's been fantastic to see the level of engagement and you've really helped sharpen the RFC for implementing the new backoffice. It's a good foundation for the work ahead of us. 

You'll hear a lot more about the project in the coming months!
 
#33
